### PR TITLE
Add module to calculate levenshtein distance between two strings

### PR DESCRIPTION
**Pull request by amishas157**
Resolves https://github.com/amishas157/osm-fudge/issues/2

### CODE REVIEW

**Review by daniel-j-h — COMMENTED** [review]
I think it would be great to have some tests here and then also testing some edge cases. For example what happens if one string if empty, both are empty. What happens with unicode symbols? :)

File: osm-fudge/levenshtein.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+    Returns:
+      The minimum number of single-character edits (insertions, deletions or substitutions)
+      required source string to  destination string
+    '''
+
+    m = len(str1)
+    n = len(str2)
+
+    d = [[0 for j in range(n + 1)] for i in range(m + 1)]
```
Comment (daniel-j-h) [inline osm-fudge/levenshtein.py]: What you could do here (if you want to): look into numpy's ndarray's and use that instead of lists of lists here :)

What you are doing is fine, though!

File: osm-fudge/levenshtein.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+                cost = 0
+            else:
+                cost = 1
+            d[i+1][j+1] = min([d[i][j+ 1] + 1, d[i+1][j] + 1, d[i][j] + cost])
+
+    return(d[m][n])
+
+
+if __name__ == "__main__":
+    distance = find_levenshtein_distance(sys.argv[1], sys.argv[2])
```
Comment (daniel-j-h) [inline osm-fudge/levenshtein.py]: Either needs an explicit check to make sure the arguments are there; or use argparser.

File: osm-fudge/n_gram.py [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+More: https://en.wikipedia.org/wiki/N-gram
+'''
+
+import sys
+
+def find_n_grams(n, str):
+    return [str[i:i+n] for i in range(len(str) - (n-1))]
+
+if __name__ == "__main__":
+    n_grams = find_n_grams(int(sys.argv[1]), sys.argv[2])
```
Comment (daniel-j-h) [inline osm-fudge/n_gram.py]: Same here

File: osm-fudge/n_gram.py [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+
+Find n-grams of a given string
+
+More: https://en.wikipedia.org/wiki/N-gram
+'''
+
+import sys
+
+def find_n_grams(n, str):
+    return [str[i:i+n] for i in range(len(str) - (n-1))]
```
Comment (daniel-j-h) [inline osm-fudge/n_gram.py]: What happens if the user passes in a negative n? What about zero? Which n make sense? Can we state that as pre-condition (assert it)?

Comment (daniel-j-h) [reply osm-fudge/n_gram.py]: How many n-grams are in a string of length l? Can we state that as a post-condition (assert it)?

Comment (amishas157) [reply osm-fudge/n_gram.py]: Couldn't get good documents for preconditon/postcondition as decorators. Did it with an assert statement

```
assert n > 0 and n <= len(str)
```

Comment (daniel-j-h) [reply osm-fudge/n_gram.py]: Yep that's exactly what I meant, perfect! :bow:

File: osm-fudge/n_gram.py [hunk 5]
```diff
@@ -8,11 +8,12 @@
 import argparse
 
 def find_n_grams(n, str):
-    return [str[i:i+n] for i in range(len(str) - (n-1))]
+	assert n > 0 and n <= len(str)
```
Comment (daniel-j-h) [inline osm-fudge/n_gram.py]: Hm do you think the check for `n <= len(str)` makes sense? What do you expect for bi-grams when the input is an empty string? Or a string with a single character? At the moment we assert in these cases. Should it be an empty list instead?

Comment (amishas157) [reply osm-fudge/n_gram.py]: Yes, thought about it. I agree that bigrams for empty string / 1-character string will not exist but it is not the case that these values are incorrect input to the function, therefore should be added as post condition. 👍

**Comment by amishas157** [comment]
Merging now. If any improvements, will do as part of new PR.
